cmd/admission-controller/main.go

```diff
@@ -68,7 +92,54 @@ func main() {
 		glog.Fatalf("failed to get kubernetes Clientset: %v", err)
 	}
 
-	webhookServer := webhook.NewWebHookServer(kubeCli, cli, certFile, keyFile)
+	kubeInformerFactory := kubeinformers.NewSharedInformerFactory(kubeCli, controller.ResyncDuration)
+
+	rl := resourcelock.EndpointsLock{
+		EndpointsMeta: metav1.ObjectMeta{
+			Namespace: ns,
+			Name:      "admission-controller",
+		},
+		Client: kubeCli.CoreV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity:      hostName,
+			EventRecorder: &record.FakeRecorder{},
+		},
+	}
+
+	webhookServer := webhook.NewWebHookServer(kubeCli, cli, kubeInformerFactory, certFile, keyFile)
+	controllerCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	kubeInformerFactory.Start(controllerCtx.Done())
+	for v, synced := range kubeInformerFactory.WaitForCacheSync(wait.NeverStop) {
+		if !synced {
+			glog.Fatalf("error syncing informer for %v", v)
+		}
+	}
+	glog.Infof("cache of informer factories sync successfully")
+
+	onStarted := func(ctx context.Context) {
+		if err := webhookServer.Run(); err != nil {
+			glog.Fatalf("stop http server %v", err)
+		}
+	}
+
+	onStopped := func() {
+		glog.Fatalf("leader election lost")
+	}
+
+	go wait.Forever(func() {
+		leaderelection.RunOrDie(controllerCtx, leaderelection.LeaderElectionConfig{
+			Lock:          &rl,
+			LeaseDuration: leaseDuration,
+			RenewDeadline: renewDuration,
+			RetryPeriod:   retryPeriod,
+			Callbacks: leaderelection.LeaderCallbacks{
+				OnStartedLeading: onStarted,
+				OnStoppedLeading: onStopped,
+			},
+		})
+	}, waitDuration)
```

Comment:
is the admission controller stateful or stateless? if it's a stateless controller, I think we don't need leader election, all replicas should be able to process requests.

Reply:
It's a stateless controller. Thx for pointing this.